Account for allocated heap memory which is unreachable from globals

mrstanb opened the pull request:
This PR:
* Adds a refinement to the reporting for `valid-memtrack`
* We now warn whenever there's allocated heap memory at program exit which is not reachable from the global program vars

Some notes:
* The SV-COMP rule for `valid-memtrack` states: `All allocated memory is tracked, i.e., pointed to or deallocated`. Hence, for now I've kept the setting of the global SV-COMP flag for `InvalidMemTrack` also for the case of not calling `free` on allocated heap memory (not necessarily only referenced by global vars). I'm open for a discussion here if you think we could improve on that as well
* There are some false positives for heap memory, reachable through local variables in case it's not freed (I also described this in test `76/09`):
```c
int main(int argc, char const *argv[]) {
    g = malloc(sizeof(int));
    g = NULL;
    int *p = malloc(sizeof(int));

    return 0; //WARN
}
```

In the above example, `p`'s memory is leaked (missing `free`), but it's also reported as unreachable (because it's not reachable via the globals, but only via a local var). Should we try to improve on that (maybe at least in the error reporting for the unreachable-from-globals-memory case) or are we ok with it as is?

Sidenote:
* Added `76/08` and `76/09` as new test cases

TODOs:
* [ ] If `earlyglobs` is activated, then abort the analysis

michael-schwarz commented:
Thank you for the PR!

> In the above example, p's memory is leaked (missing free), but it's also reported as unreachable (because it's not reachable via the globals, but only via a local var).  Should we try to improve on that (maybe at least in the error reporting for the unreachable-from-globals-memory case) or are we ok with it as is?

Confusingly this is exactly the definition of `valid-memtrack` in SV-COMP (I was also surprised by this). See e.g.: https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1464
So we cannot really improve here and actually need to report this as a violation of `valid-memtrack`.

michael-schwarz commented on tests/regression/76-memleak/09-unreachable-with-local-var.c:
```diff
@@ -0,0 +1,17 @@
+    g = malloc(sizeof(int));
+    // Reference to g's heap contents is lost here
+    g = NULL;
+    // We get a false positive for p's memory being unreachable
+    // It's still leaked, but due to free() being commented out
+    // TODO: Should we only improve the error reporting for unreachable memory in this case?
```
At least for `valid-memtrack` this is not actually a false-positive.

mrstanb replied:
Ah, yes. I saw from the discussion in https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1464 that if the memory is reachable from locals then it's a correct reporting of a `valid-memtrack` violation.

I'll remove the `// TODO` in this case.

michael-schwarz commented on src/analyses/memLeak.ml:
```diff
@@ -19,6 +19,22 @@ struct
+    (* Filtering by GVar seems to account for declarations, as well as definitions of global vars *)
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
+
+  let get_reachable_mem_from_globals (global_vars:varinfo list) ctx =
+    global_vars
+    |> List.map (fun v -> Lval (Var v, NoOffset))
+    |> List.filter_map (fun exp ->
```
This can be improved a little: If we have global structs for example, we should follow all the pointer fields of the struct.

mrstanb replied:
Thanks for the idea! I quickly tried out a few regr. tests with global struct variables (both pointer and non-pointer ones, as well as with nested structs; cf. `76/{10, 11, 12}`)

It seems that without changing anything in the implementation, Goblint can already detect the `valid-memtrack` violations in these test cases.

Nonetheless, if you think it makes sense to add anything further to the implementation, I'm more than happy to do that.

michael-schwarz replied:
> It seems that without changing anything in the implementation, Goblint can already detect the `valid-memtrack` violations in these test cases.

Great! I think the other interesting case would be the cases where `valid-memtrack` holds, because something is still reachable via a global struct member. Have you also tried that direction?

sim642 replied:
If it means reachable, then it should use `Queries.ReachableFrom`.

michael-schwarz replied:
No, it is **must**-reachable, while the query is **may**-reachable.

sim642 replied:
The current `MayPointTo` query is also a may-query. The must-ness comes from the singleton check of the may result.

Maybe the non-transitive aspect of `MayPointTo` is fine here and must-reachability via pointers would just make us less precise? In that case the function should also be named accordingly to avoid the inconsistency.

michael-schwarz replied:
Must-reachability by pointers would make us more precise: One would consider all the things that the globals (or it's struct members) must point to (i.e. singleton points-to-set), and then can consider these as roots as well to include such things that must-be-reachable from these.
Doing a MayReachable query and then only considering the set if it is a singleton set would be less precise than considering the MayPointTo set if it is singleton.

mrstanb replied:
> Great! I think the other interesting case would be the cases where valid-memtrack holds, because something is still (edit: must-)reachable via a global struct member. Have you also tried that direction?

I added some changes with respect to this suggestion. Please check out my comment here: https://github.com/goblint/analyzer/pull/1241#issuecomment-1813364110

michael-schwarz commented on src/analyses/memLeak.ml:
```diff
@@ -27,17 +43,25 @@ struct
+      (* Check and warn if some of the allocated memory is not deallocated at program exit *)
       match assert_exp_imprecise, exp with
       | true, Some exp ->
         set_mem_safety_flag InvalidMemTrack;
```
We are already setting that flag above now.

```suggestion
```

michael-schwarz commented on src/analyses/memLeak.ml:
```diff
@@ -27,17 +43,25 @@ struct
-        M.warn ~category:(Behavior (Undefined MemoryLeak)) ~tags:[CWE 401] "assert expression %a is unknown. Memory leak might possibly occur for heap variables: %a" d_exp exp D.pretty state
+        M.warn ~category:(Behavior (Undefined MemoryLeak)) ~tags:[CWE 401] "Assert expression %a is unknown. Memory leak might possibly occur for heap variables: %a" d_exp exp D.pretty allocated_mem
       | _ ->
         set_mem_safety_flag InvalidMemTrack;
```
Same here.

```suggestion
```

michael-schwarz commented:
I will start a run with the two suggestions from above applied to see what the results are.

michael-schwarz commented:
With these two suggestions on top of #1234, we succeed on an additional 17 tasks:

```
Statistics:           3862 Files
  correct:            1175
    correct true:     1175
    correct false:       0
  incorrect:             0
    incorrect true:      0
    incorrect false:     0
  unknown:            2687
  Score:              2350 (max: 5857)
```

michael-schwarz commented:
Tables: [results.2023-11-08_10-02-03.zip](https://github.com/goblint/analyzer/files/13294538/results.2023-11-08_10-02-03.zip)

mrstanb commented:
I added some changes regarding the check of whether some memory is still transitively reachable via the points-to set of a global struct variable (both of pointer and non-pointer type), as well as its fields and the contents of the points-to sets.

I also added two new regression test cases to check if the changes make sense (cf. `76/{13, 14}`).

mrstanb commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@
+            | Queries.VD.Struct s ->
+              let struct_fields = ValueDomain.Structs.keys s in
+              let ptr_struct_fields = List.filter (fun f -> isPointerType f.ftype) struct_fields in
+              if List.length ptr_struct_fields = 0 then None else Some (s, ptr_struct_fields)
```
Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/845910410bbda2f9964ccd727d63a6e7b811ac7f

mrstanb commented:
Maybe we should add a check that if `earlyglobs` is activated, then we should abort the analysis?

Iirc, we discussed with @jerhard that `earlyglobs` would stop us from obtaining meaningful information about what global vars can reference.

sim642 commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@ struct
   let context _ d = d
 
   (* HELPER FUNCTIONS *)
+  let get_global_vars () =
+    (* Filtering by GVar seems to account for declarations, as well as definitions of global vars *)
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
```
Does CIL really guarantee that? I think in other places we don't assume it does and also handle `GVarDecl`.

mrstanb replied:
I did some manual inspection to check for that. Iirc, `GVarDecl` was giving me a bunch of stuff from the header files as well which seemed a bit too much for me. Anyhow, I can add `GVarDecl` in there as well if we don't have a CIL guarantee regarding `GVar`.

mrstanb replied:
Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/6cc01b5177b6bc31899c290b6ae65660bf4aa805

sim642 commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@ struct
+    (* Filtering by GVar seems to account for declarations, as well as definitions of global vars *)
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
+
+  let get_global_struct_ptr_vars () =
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
+    |> List.filter (fun v -> match v.vtype with TPtr (TComp _, _) | TPtr ((TNamed ({ttype = TComp _; _}, _)), _) -> true | _ -> false)
+
+  let get_global_struct_non_ptr_vars () =
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
+    |> List.filter (fun v -> match v.vtype with TComp (_, _) | (TNamed ({ttype = TComp _; _}, _)) -> true | _ -> false)
```
Wouldn't `List.partition` give both in a single pass?

mrstanb replied:
I'll try it out, thanks for the tip!

mrstanb replied:
I'm only not sure about the predicate formulation for `List.partition`, namely: we'd have the `true` case for the `TPtr` structs and the `false` one for the structs without a `TPtr` type. In this case what should we do with the rest of the types (otherwise, the match wouldn't be complete)?

michael-schwarz replied:
Yes, then it is not quite a good usecase for `partition`. We could probably come up with an efficient thing that splits it into two lists and discards the rest in linear time, but I don't think this is crucial for this year's sv-comp.

sim642 commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@ struct
+
+  let get_global_struct_ptr_vars () =
+    List.filter_map (function GVar (v, _, _) -> Some v | _ -> None) !Cilfacade.current_file.globals
+    |> List.filter (fun v -> match v.vtype with TPtr (TComp _, _) | TPtr ((TNamed ({ttype = TComp _; _}, _)), _) -> true | _ -> false)
```
This isn't very reliable, there may be more than one `typedef`. To do it properly, matching should be on `Cil.unrollType`.

mrstanb replied:
Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/6cc01b5177b6bc31899c290b6ae65660bf4aa805

sim642 commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@ struct
+                  begin match ValueDomain.Structs.get s f with
+                    | Queries.VD.Address a ->
+                      let reachable_from_addr_set =
+                        List.fold_left (fun acc addr ->
+                            match addr with
+                            | Queries.AD.Addr.Addr (v, _) -> List.append acc (v :: get_reachable_mem_from_str_ptr_globals [v] ctx)
+                            | _ -> acc
+                          ) [] (Queries.AD.elements a)
```
No need for `List`, just `Queries.AD.fold`.

mrstanb replied:
Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/6cc01b5177b6bc31899c290b6ae65660bf4aa805

sim642 commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@ struct
+                      let reachable_from_addr_set =
+                        List.fold_left (fun acc addr ->
+                            match addr with
+                            | Queries.AD.Addr.Addr (v, _) -> List.append acc (v :: get_reachable_mem_from_str_ptr_globals [v] ctx)
```
Appending to the end of `acc` isn't very efficient. It's better to prepend new things.

This occurs elsewhere as well.

mrstanb replied:
Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/6cc01b5177b6bc31899c290b6ae65660bf4aa805

michael-schwarz commented:
> Maybe we should add a check that if earlyglobs is activated, then we should abort the analysis?

I don't think we use `earlyglobs` in sv-comp, so this doesn't seem too pressing.

michael-schwarz commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,105 @@ struct
+                  begin match ValueDomain.Structs.get s f with
+                    | Queries.VD.Address a ->
+                      let reachable_from_addr_set =
+                        List.fold_left (fun acc addr ->
```
I think we would need to check that this is singleton here once again. If it **may** point to several things, none of them are **must**-reachable.

mrstanb replied:
Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/80492ccd1dcad21e49a949a989d1cee42bbb6585. I also added an `is_top` check that was missing for the address set `a`.

michael-schwarz commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,111 @@ struct
+
+  let get_global_struct_non_ptr_vars () =
+    get_global_vars ()
+    |> List.filter (fun v -> match v.vtype with TComp (_, _) | (TNamed ({ttype = TComp _; _}, _)) -> true | _ -> false)
```
Probably we should also call `unrollType` here?

mrstanb replied:
Seems that I've forgotten to add it here. Should be fixed now: https://github.com/goblint/analyzer/pull/1241/commits/f2ca6d146e1271709259be24d5b20f9e61aab7dd

michael-schwarz commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,111 @@ struct
+    in
+    global_struct_ptr_vars
+    |> List.fold_left (fun acc var ->
+        if ctx.ask (Queries.IsHeapVar var) then eval_value_of_heap_var var
+        else if not (ctx.ask (Queries.IsAllocVar var)) && isPointerType var.vtype then get_pts_of_non_heap_ptr_var var
+        else acc
```
Is it intended that we drop `acc` here in the first two cases?

mrstanb replied:
No, I've simply forgotten to prepend the returned `varinfo list`s in the first two cases to `acc`. Should be fixed in https://github.com/goblint/analyzer/pull/1241/commits/0e09d099d5facf7354014319805eb53e186119d2

sim642 commented on src/analyses/memLeak.ml:
```diff
@@ -20,6 +20,115 @@ struct
+              match ValueDomain.Structs.get s field with
+              | Queries.VD.Address a ->
+                let reachable_from_addr_set =
+                  List.fold_left (fun acc_addr addr ->
+                      match addr with
+                      | Queries.AD.Addr.Addr (v, _) -> (v :: get_reachable_mem_from_str_ptr_globals [v] ctx) @ acc_addr
+                      | _ -> acc_addr
+                    ) [] (Queries.AD.elements a)
```
This still involves an unnecessary intermediate list.

mrstanb replied:
Sorry, should be now refactored correctly in https://github.com/goblint/analyzer/pull/1241/commits/9153eb3becd5904b99aa8250e50b2cd22f74a128

mrstanb commented:
> I may have missed this, but in what sense are global structs special? As in, why aren't we also considering the contents of global arrays and whatnot?
Or are we just less precise for global array contents then?

I'm not really sure about arrays, because I haven't worked a lot with the array domain of Goblint and so I can't express an opinion on that.

I've thought about global `union`s, however I'm unsure about what should happen in a situation like the following. Consider:
```c
typedef union {
    int i;
    char *str;
} un;

un *glob_var;

int main() {
    ...
}
```

What happens if the program first accesses `str` of `glob_var` and then subsequently accesses `i` of `glob_var`? Wouldn't `str` be "lost" so to speak? I mean the memory is there, but `i` is the thing that's now occupying the place of `str`, unless I'm thinking in the wrong direction.

michael-schwarz commented:
Yeah, unions do not seem to be a good candidate here. We discussed the issue of arrays at our Gobcon yesterday, and decided that due to the abstraction we use for arrays, it is unlikely that things are must-reachable from arrays and have thus decided against implementing dedicated support for it for now.

michael-schwarz commented:
I think we can now merge this into master as well? Wdyt @sim642?
